File: java/net/mikekohn/java_grinder/Atari2600.java
/**
 *  Java Grinder
 *  Author: Michael Kohn
 *   Email: mike@mikekohn.net
 *     Web: http://www.naken.cc/
 * License: GPL
 *
 * Copyright 2014-2015 by Michael Kohn
 *
 */

package net.mikekohn.java_grinder;

abstract public class Atari2600
{
  protected Atari2600()
  {
  }

  public static final int PLAYFIELD_BALL_SIZE_8 = 0x30;
  public static final int PLAYFIELD_BALL_SIZE_4 = 0x20;
  public static final int PLAYFIELD_BALL_SIZE_2 = 0x10;
  public static final int PLAYFIELD_BALL_SIZE_1 = 0x00;
  public static final int PLAYFIELD_HIGH_PRIORITY = 0x04;
  public static final int PLAYFIELD_SCORE_COLOR = 0x02;
  public static final int PLAYFIELD_REFLECT = 0x01;

  public static final int SWITCH_RESET = 0x01;
  public static final int SWITCH_SELECT = 0x02;
  public static final int SWITCH_COLOR = 0x08;
  public static final int SWITCH_DIFFICULTY_0 = 0x40;
  public static final int SWITCH_DIFFICULTY_1 = 0x80;

  public static void waitHsync(int lines) { }
  public static void waitHsync() { }
  public static void startVblank() { }
  public static void waitVblank() { }
  public static void startOverscan() { }
  public static void waitOverscan() { }
  public static void setColorPlayer0(int color) { }
  public static void setColorPlayer1(int color) { }
  public static void setColorPlayfield(int color) { }
  public static void setColorBackground(int color) { }

  public static void resetPlayer0() { }
  public static void resetPlayer1() { }
  public static void resetMissile0() { }
  public static void resetMissile1() { }
  public static void resetBall() { }
  public static void resetMissile0ToPlayer0On() { }
  public static void resetMissile1ToPlayer1On() { }
  public static void resetMissile0ToPlayer0Off() { }
  public static void resetMissile1ToPlayer1Off() { }
  public static void applyHorizontalMotion() { }
  public static void clearMotionRegisters() { }
  public static void clearCollisionLatches() { }

  public static void enableMissile0() { }
  public static void enableMissile1() { }
  public static void enableBall() { }
  public static void disableMissile0() { }
  public static void disableMissile1() { }
  public static void disableBall() { }

  public static void enablePlayer0Reflect() { }
  public static void enablePlayer1Reflect() { }
  public static void disablePlayer0Reflect() { }
  public static void disablePlayer1Reflect() { }

  public static void setPlayfieldMode(int mode) { }

  public static void setPlayfield0(byte value) { }
  public static void setPlayfield1(byte value) { }
  public static void setPlayfield2(byte value) { }

  public static void setPlayfieldData(byte[] data) { }
  public static void setPlayfieldLength(byte value) { }

  public static void setPlayer0Position(byte x, byte y) { }
  public static void setPlayer1Position(byte x, byte y) { }
  public static void setMissile0Position(byte x, byte y) { }
  public static void setMissile1Position(byte x, byte y) { }
  public static void setBallPosition(byte x, byte y) { }

  public static void setPlayer0Sprite(byte[] data) { }
  public static void setPlayer1Sprite(byte[] data) { }
  public static void setMissile0Sprite(byte[] data) { }
  public static void setMissile1Sprite(byte[] data) { }
  public static void setBallSprite(byte[] data) { }

  public static void drawScreen() { }

  public static boolean isCollisionMissile0Player1() { return false; }
  public static boolean isCollisionMissile0Player0() { return false; }
  public static boolean isCollisionMissile1Player0() { return false; }
  public static boolean isCollisionMissile1Player1() { return false; }
  public static boolean isCollisionPlayer0PlayField() { return false; }
  public static boolean isCollisionPlayer0Ball() { return false; }
  public static boolean isCollisionPlayer1PlayField() { return false; }
  public static boolean isCollisionPlayer1Ball() { return false; }
  public static boolean isCollisionMissile0Playfield() { return false; }
  public static boolean isCollisionMissile0Ball() { return false; }
  public static boolean isCollisionMissile1Playfield() { return false; }
  public static boolean isCollisionMissile1Ball() { return false; }
  public static boolean isCollisionBallPlayfield() { return false; }
  public static boolean isCollisionPlayer0Player1() { return false; }
  public static boolean isCollisionMissile0Missile1() { return false; }

  /** Audio sound type, 4 bit number */
  public static void setAudioControl0(byte value) { }

  /** Audio sound type, 4 bit number */
  public static void setAudioControl1(byte value) { }

  /** Audio frequency, 5 bit number */
  public static void setAudioFrequency0(byte value) { }

  /** Audio frequency, 5 bit number */
  public static void setAudioFrequency1(byte value) { }

  /** Audio volume, 4 bit number */
  public static void setAudioVolume0(byte value) { }

  /** Audio volume, 4 bit number */
  public static void setAudioVolume1(byte value) { }

  public static boolean isJoystick0Right() { return false; }
  public static boolean isJoystick0Left() { return false; }
  public static boolean isJoystick0Down() { return false; }
  public static boolean isJoystick0Up() { return false; }
  public static boolean isJoystick0ButtonDown() { return false; }
  public static boolean isJoystick1Right() { return false; }
  public static boolean isJoystick1Left() { return false; }
  public static boolean isJoystick1Down() { return false; }
  public static boolean isJoystick1Up() { return false; }
  public static boolean isJoystick1ButtonDown() { return false; }

  public static byte getSwitches() { return -1; }
  public static boolean isSwitchResetDown() { return false; }
  public static boolean isSwitchSelectDown() { return false; }
  public static boolean isSwitchColorSet() { return false; }
  public static boolean isSwitchDifficultyP0Down() { return false; }
  public static boolean isSwitchDifficultyP1Down() { return false; }

  public static void setScore0(byte value) { }
  public static void setScore1(byte value) { }

  public static void setTitlePos(int value) { }
  public static void setTitleColor(int value) { }
  public static void drawTitleScreen() { }

  /** Change ROM bank.  Must be a constant and requires concatinating
      more than 1 bin file together */
  public static void setBank(byte index) { }
}

File: api/atari_2600.cxx
/**
 *  Java Grinder
 *  Author: Michael Kohn
 *   Email: mike@mikekohn.net
 *     Web: http://www.mikekohn.net/
 * License: GPLv3
 *
 * Copyright 2014-2021 by Michael Kohn
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

#include "api/joystick.h"

#define CHECK_FUNC(funct,sig) \
  if (strcmp(#funct#sig, method_name) == 0) \
  { \
    return generator->atari2600_##funct##sig(); \
  }

#define CHECK_FUNC_CONST(funct,sig) \
  if (strcmp(#funct#sig, method_name) == 0) \
  { \
    return generator->atari2600_##funct##sig(const_val); \
  }

#define CHECK_FUNC_CONST_2(funct,sig) \
  if (strcmp(#funct#sig, function) == 0) \
  { \
    return generator->atari2600_##funct##sig(const_val1, const_val2); \
  }

int atari_2600(JavaClass *java_class, Generator *generator, const char *method_name)
{
  CHECK_FUNC(waitHsync,_I)
  CHECK_FUNC(waitHsync,)
  CHECK_FUNC(startVblank,)
  CHECK_FUNC(waitVblank,)
  CHECK_FUNC(startOverscan,)
  CHECK_FUNC(waitOverscan,)
  CHECK_FUNC(setColorPlayer0,_I)
  CHECK_FUNC(setColorPlayer1,_I)
  CHECK_FUNC(setColorPlayfield,_I)
  CHECK_FUNC(setColorBackground,_I)

  CHECK_FUNC(resetPlayer0,)
  CHECK_FUNC(resetPlayer1,)
  CHECK_FUNC(resetMissile0,)
  CHECK_FUNC(resetMissile1,)
  CHECK_FUNC(resetBall,)
  CHECK_FUNC(resetMissile0ToPlayer0On,)
  CHECK_FUNC(resetMissile1ToPlayer1On,)
  CHECK_FUNC(resetMissile0ToPlayer0Off,)
  CHECK_FUNC(resetMissile1ToPlayer1Off,)
  CHECK_FUNC(applyHorizontalMotion,)
  CHECK_FUNC(clearMotionRegisters,)
  CHECK_FUNC(clearCollisionLatches,)

  CHECK_FUNC(enableMissile0,)
  CHECK_FUNC(enableMissile1,)
  CHECK_FUNC(enableBall,)
  CHECK_FUNC(disableMissile0,)
  CHECK_FUNC(disableMissile1,)
  CHECK_FUNC(disableBall,)

  CHECK_FUNC(enablePlayer0Reflect,)
  CHECK_FUNC(enablePlayer1Reflect,)
  CHECK_FUNC(disablePlayer0Reflect,)
  CHECK_FUNC(disablePlayer1Reflect,)

  CHECK_FUNC(setPlayfieldMode,_I)

  CHECK_FUNC(setPlayfield0,_B)
  CHECK_FUNC(setPlayfield1,_B)
  CHECK_FUNC(setPlayfield2,_B)

  CHECK_FUNC(setPlayfieldData,_aB)
  CHECK_FUNC(setPlayfieldLength,_B)

  CHECK_FUNC(setPlayer0Position,_BB)
  CHECK_FUNC(setPlayer1Position,_BB)
  CHECK_FUNC(setMissile0Position,_BB)
  CHECK_FUNC(setMissile1Position,_BB)
  CHECK_FUNC(setBallPosition,_BB)

  CHECK_FUNC(setPlayer0Sprite,_aB)
  CHECK_FUNC(setPlayer1Sprite,_aB)
  CHECK_FUNC(setMissile0Sprite,_aB)
  CHECK_FUNC(setMissile1Sprite,_aB)
  CHECK_FUNC(setBallSprite,_aB)

  CHECK_FUNC(drawScreen,)

  CHECK_FUNC(isCollisionMissile0Player1,)
  CHECK_FUNC(isCollisionMissile0Player0,)
  CHECK_FUNC(isCollisionMissile1Player0,)
  CHECK_FUNC(isCollisionMissile1Player1,)
  CHECK_FUNC(isCollisionPlayer0PlayField,)
  CHECK_FUNC(isCollisionPlayer0Ball,)
  CHECK_FUNC(isCollisionPlayer1PlayField,)
  CHECK_FUNC(isCollisionPlayer1Ball,)
  CHECK_FUNC(isCollisionMissile0Playfield,)
  CHECK_FUNC(isCollisionMissile0Ball,)
  CHECK_FUNC(isCollisionMissile1Playfield,)
  CHECK_FUNC(isCollisionMissile1Ball,)
  CHECK_FUNC(isCollisionBallPlayfield,)
  CHECK_FUNC(isCollisionPlayer0Player1,)
  CHECK_FUNC(isCollisionMissile0Missile1,)

  CHECK_FUNC(setAudioControl0,_B)
  CHECK_FUNC(setAudioControl1,_B)
  CHECK_FUNC(setAudioFrequency0,_B)
  CHECK_FUNC(setAudioFrequency1,_B)
  CHECK_FUNC(setAudioVolume0,_B)
  CHECK_FUNC(setAudioVolume1,_B)

  CHECK_FUNC(getSwitches,)
  CHECK_FUNC(isSwitchResetDown,)
  CHECK_FUNC(isSwitchSelectDown,)
  CHECK_FUNC(isSwitchColorSet,)
  CHECK_FUNC(isSwitchDifficultyP0Down,)
  CHECK_FUNC(isSwitchDifficultyP1Down,)

  CHECK_FUNC(setScore0,_B)
  CHECK_FUNC(setScore1,_B)

  CHECK_FUNC(setTitlePos,_I)
  CHECK_FUNC(setTitleColor,_I)
  CHECK_FUNC(drawTitleScreen,)

  CHECK_FUNC(setBank,_B)

  return -1;
}

int atari_2600(JavaClass *java_class, Generator *generator, const char *method_name, int const_val)
{
  CHECK_FUNC_CONST(waitHsync,_I)

  CHECK_FUNC_CONST(setColorPlayer0,_I)
  CHECK_FUNC_CONST(setColorPlayer1,_I)
  CHECK_FUNC_CONST(setColorPlayfield,_I)
  CHECK_FUNC_CONST(setColorBackground,_I)

  CHECK_FUNC_CONST(setPlayfieldMode,_I)

  CHECK_FUNC_CONST(setPlayfield0,_B)
  CHECK_FUNC_CONST(setPlayfield1,_B)
  CHECK_FUNC_CONST(setPlayfield2,_B)

  CHECK_FUNC_CONST(setAudioControl0,_B)
  CHECK_FUNC_CONST(setAudioControl1,_B)
  CHECK_FUNC_CONST(setAudioFrequency0,_B)
  CHECK_FUNC_CONST(setAudioFrequency1,_B)
  CHECK_FUNC_CONST(setAudioVolume0,_B)
  CHECK_FUNC_CONST(setAudioVolume1,_B)
  CHECK_FUNC_CONST(setBank,_B)

  return -1;
}

int atari_2600(JavaClass *java_class, Generator *generator, const char *function, int const_val1, int const_val2)
{
  //CHECK_FUNC_CONST_2(someFunction,_II)

  return -1;
}

